File: shared/src/types/task/graph.ts
import type { Task } from './schema'

/**
 * The order a run walks the graph: layers of node ids, triggers first, every node after
 * everything that feeds it. Null when the graph has a cycle — the editor refuses the edge
 * and the server refuses the run with the same answer.
 */
export function runOrder(task: Task): string[][] | null {
  const waiting = new Map(task.nodes.map((one) => [one.id, 0]))
  for (const edge of task.edges) waiting.set(edge.to, (waiting.get(edge.to) ?? 0) + 1)
  const layers: string[][] = []
  let ready = task.nodes.filter((one) => waiting.get(one.id) === 0).map((one) => one.id)
  let placed = 0
  while (ready.length > 0) {
    layers.push(ready)
    placed += ready.length
    const next: string[] = []
    for (const edge of task.edges) {
      if (!ready.includes(edge.from)) continue
      const left = (waiting.get(edge.to) ?? 0) - 1
      waiting.set(edge.to, left)
      if (left === 0) next.push(edge.to)
    }
    ready = next
  }
  return placed === task.nodes.length ? layers : null
}

File: daemon/internal/task/graph.go
package task

// RunOrder is the order a run walks the graph: layers of node ids, triggers first, every node
// after everything that feeds it. Nil when the graph has a cycle — the editor refuses the edge
// and the server refuses the run with the same answer.
func RunOrder(t Task) [][]string {
	waiting := make(map[string]int, len(t.Nodes))
	for _, node := range t.Nodes {
		waiting[node.ID] = 0
	}
	for _, edge := range t.Edges {
		waiting[edge.To]++
	}

	var ready []string
	for _, node := range t.Nodes {
		if waiting[node.ID] == 0 {
			ready = append(ready, node.ID)
		}
	}

	var layers [][]string
	placed := 0
	for len(ready) > 0 {
		layers = append(layers, ready)
		placed += len(ready)
		leaving := map[string]bool{}
		for _, id := range ready {
			leaving[id] = true
		}
		var next []string
		for _, edge := range t.Edges {
			if !leaving[edge.From] {
				continue
			}
			waiting[edge.To]--
			if waiting[edge.To] == 0 {
				next = append(next, edge.To)
			}
		}
		ready = next
	}
	if placed != len(t.Nodes) {
		return nil
	}
	return layers
}
